// @ngneat/reactive-forms/esm2015/lib/formBuilder.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { FormBuilder as NgFormBuilder } from '@angular/forms';
import { FormArray } from './formArray';
import { FormControl } from './formControl';
import { FormGroup } from './formGroup';
import * as i0 from "@angular/core";
function isAbstractControlOptions(options) {
    return (options.asyncValidators !== undefined ||
        options.validators !== undefined ||
        options.updateOn !== undefined);
}
let FormBuilder = class FormBuilder extends NgFormBuilder {
    group(controlsConfig, options) {
        const controls = this._reduceControls(controlsConfig);
        let validators = null;
        let asyncValidators = null;
        let updateOn;
        if (options != null) {
            if (isAbstractControlOptions(options)) {
                validators = options.validators != null ? options.validators : null;
                asyncValidators = options.asyncValidators != null ? options.asyncValidators : null;
                updateOn = options.updateOn != null ? options.updateOn : undefined;
            }
            else {
                // `options` are legacy form group options
                validators = options['validator'] != null ? options['validator'] : null;
                asyncValidators = options['asyncValidator'] != null ? options['asyncValidator'] : null;
            }
        }
        // Todo: dan remove the any
        return new FormGroup(controls, { asyncValidators, updateOn, validators });
    }
    control(formState, validatorOrOpts, asyncValidator) {
        return new FormControl(formState, validatorOrOpts, asyncValidator);
    }
    array(controlsConfig, validatorOrOpts, asyncValidator) {
        const controls = controlsConfig.map(c => this._createControl(c));
        return new FormArray(controls, validatorOrOpts, asyncValidator);
    }
};
FormBuilder.ɵprov = i0.ɵɵdefineInjectable({ factory: function FormBuilder_Factory() { return new FormBuilder(); }, token: FormBuilder, providedIn: "root" });
FormBuilder = __decorate([
    Injectable({ providedIn: 'root' })
], FormBuilder);
export { FormBuilder };
